File: leetcode/medium/932_beautiful_array.py
# -----------------------------------------
# Model Solution
#
# Time  Complexity: O(nlog(n))
# Space Complexity: O(nlog(n))
# -----------------------------------------
# Ref:
# a) https://leetcode.com/problems/beautiful-array/solution/
# b) https://www.cnblogs.com/grandyang/p/12287146.html
from functools import lru_cache

class Solution:
    def beautifulArray(self, N: int) -> List[int]:
        @lru_cache(None)
        def beautiful_array(n):
            if n == 1:
                return [1]
            odds  = beautiful_array((n + 1) // 2)
            evens = beautiful_array(n // 2)
            return [2 * num - 1 for num in odds] + [2 * num for num in evens]

        return beautiful_array(N)

File: leetcode/medium/932_beautiful_array.cpp
// -----------------------------------------
// Model Solution
//
// Time  Complexity: O(n)
// Space Complexity: O(n)
// -----------------------------------------
// Ref: https://leetcode.com/problems/beautiful-array/discuss/186679/Odd-%2B-Even-Pattern-O(N)
class Solution {
public:
    vector<int> beautifulArray(int n) {
        vector<int> beautiful_array = {1};
        while (beautiful_array.size() < n) {
            vector<int> tmp_array;

            for (int& num: beautiful_array) {
                if (num * 2 <= n)
                    tmp_array.push_back(num * 2);
            }

            for (int& num: beautiful_array) {
                if (num * 2 - 1 <= n)
                    tmp_array.push_back(num * 2 - 1);
            }

            beautiful_array = tmp_array;
        }

        return beautiful_array;
    }
};
